### ISSUE TITLE
Add support for loading and saving AVIF images

### ISSUE THREAD

**Issue by YakoYakoYokuYoku** [godotengine/godot-proposals#2461]
### Describe the project you are working on

With the engine itself.

### Describe the problem or limitation you are having in your project

As AVIF, a format based on AV1 is getting adopted widely in Firefox and Chromium based browsers (still in experimental phase) and in modern software like ImageMagick, GIMP and Krita, Godot could also see some support for it.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

AVIF is characterized by being an open standard, its compression efficiency, having HDR and chroma subsampling support.
Both [`libheif`](https://github.com/strukturag/libheif) and [`libavif`](https://github.com/AOMediaCodec/libavif) bring reading/writing for the format. Encoding is done by `rav1e` and `dav1d` does the decoding, otherwise `libaom` can do both.
I've also considered requesting for HEIF support, but due to `x265`'s GPL-2.0 licensing this may not be possible.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Basically just like the usage of images in a Godot project, i.e. adding a texture to a 2D sprite, dragging images to the 2D editor, setting the project icon, setting a custom mouse, using AVIF images inside the scripts, etc.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

N/A.

### Is there a reason why this should be core and not an add-on in the asset library?

Although this can be partially implemented by using an asset, problems still arise like no thumbnail support, having to resort to bitmaps in some cases or depending on the state of the asset.

**Comment by Calinou** [comment]
I thought about this a while ago, but I wonder if libavif suffers from similar downsides as libwebp (relatively large size, difficult to build for all platforms).

**Comment by YakoYakoYokuYoku** [comment]
> I thought about this a while ago, but I wonder if libavif suffers from similar downsides as libwebp (relatively large size, difficult to build for all platforms).

Well, there are three observations we can take from this, one is that the size of the libs (`libheif` and `libavif`) is added from the encoders/decoders too, two is that the sources of both of them reside in one folder respectively and three is that because we are requiring the codecs for AV1, the latter *could* also be used to bring support for the latter inside Godot as well.

> Also, one downside of AVIF over PNG is that it requires much more processing power to decode, which will in turn slow down the game's loading times (especially on mobile).

This is some of my *little* testing with AVIF I did in my machine.

``` stdout
$ # Original file https://wallpapercave.com/wp/wp4618883.jpg
$ magick wp4618883.jpg -alpha on -define png:compression-level=9 drstone2.png
$ magick wp4618883.jpg -alpha on -quality 40 drstone4.avif 

$ # Both times of converting images to BMP with `magick`
magick drstone2.png drstone2.bmp  0.25s user 0.21s system 171% cpu 0.267 total
magick drstone4.avif drstone2.bmp  0.41s user 0.12s system 95% cpu 0.558 total

$ identify drstone2.png drstone4.avif 
drstone2.png PNG 2835x2523 2835x2523+0+0 8-bit sRGB 6.59573MiB 0.000u 0:00.000
drstone4.avif AVIF 2835x2523 2835x2523+0+0 8-bit sRGB 0.010u 0:00.000

$ du -Bk drstone2.png drstone4.avif
4560K	drstone2.png  # OptiPNG may do a better job too
200K	drstone4.avif  # 76k if using 10 as quality
```

I don't know what to say about mobile (yet) but using AVIF could be very useful when dealing with large textures for very big and heavy games for example.

> We'd prefer not to introduce Rust dependencies in core Godot to avoid making the buildsystem even more complex. Therefore, we need something written in C or C++.

`libaom` would certainly be the way to go.

**Comment by YakoYakoYokuYoku** [comment]
Another alternative would be to use a `tinyavif` thirdparty module, much like [link-u/cavif](https://github.com/link-u/cavif).

**Comment by fire** [comment]
https://github.com/link-u/cavif is short in complexity, I can see it being doable.

What formats are supported by avif?

I saw hdr and 8bit support.

Are you able to provide samples of avif pictures?

I'm not promising a cavif module but it seems like the most appropriate way of getting the feature.

Are you able to work on pull request?

**Comment by fire** [comment]
Summary: game engine convert raw sources in RGBAH, RGBAF, or RGBA8 into compressed texture formats.

This will only provide yet another way of going from (example) RGBA8 -> ETC2 or DXT5 or BC7.

Decompressing raw RGBA8 on any computer is expensive.

AVIF makes the engine more accessible but we already use webp? as the internal format for lossless. So it's not a massive improvement like implementing fast bc7 compression.

**Comment by Calinou** [comment]
@fire WebP, like AVIF, would mainly be used in 2D games with large textures but where video RAM compression isn't viable due to its low quality.

**Comment by YakoYakoYokuYoku** [comment]
About AVIF's colorspace it uses YUV instead of RGB, so getting the right one depends on the decoder or the YUV to RGB converter.

> Are you able to work on pull request?

Yes I can, and I'm also doing local work. As of now AVIF 2D sprites loads correctly in the 2D scene editor.

**Comment by novomesk** [comment]
Hello,

I worked with both libheif (contributed AVIF support to GIMP) and libavif (contributed AVIF Qt plug-in to kimageformats).

Each of these libraries have their advantages and disadvantages. The choice is a matter of personal preferences.

These libraries itself are relatively small, final size depends on the list of encoders/decoders being selected.

For example when building my [Qt plug-in](https://github.com/novomesk/qt-avif-image-plugin) for Windows, I add only libaom, because it provides both reading and writing AVIF files. Final size of the plug-in with statically linked libavif and libaom is approximately 10MB.

However, author of [qView](https://github.com/jurplel/qView) needed only read functionality in his viewer, so he used only dav1d decoder. The final size of the Qt plug-in with libavif and dav1d is only 1.02 MB!!!

So if you want AVIF read-only support, increase in binary size will be approximately 1MB,
if you want read/write AVIF support, increase could be up-to 10MB.

I'd like to see AVIF support (at least for import) in Godot Engine.

**Comment by fire** [comment]
It has been a repeated concern that AVIF can be slow to encode. We didn't want to be significantly slower than png encoding.

Can you research this and comment?

**Comment by novomesk** [comment]
Export to AVIF used to be extremely slow in the past. But as AV1 encoders (libaom, rav1e, svt-av1) improve, time to save AVIF and memory usage [dropped significantly](https://web.dev/compress-images-avif/).

Libavif has 0-10 speed levels. One month ago I measured libavif 0.9.0 with libaom 3.1+ on laptop with Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz (2 cores, 4 threads with AVX2 support). I used randomly selected **3264x2448** image from mobile phone.

libavif speed 1: 1minute + 15seconds

libavif speed 5: 10,2 seconds, size of AVIF file just 1% higher.

libavif speed 6: **3,3 seconds**, 2.1% more data than AVIF at speed 1

So at speed level 6, AVIF is slightly slower on the laptop than PNG (which also needs few seconds on large images).

**Comment by YakoYakoYokuYoku** [comment]
Reading from what @novomesk is mentioning and some ideas I had in the past I've came up with the following, implementing a generic AV1 codec that uses any of the three AV1 libraries mentioned above and implementing a microparser for the AVIF files. The generic library has the advantage of providing AV1 support for the engine and the microparser will not have any strange dependencies when building the bundled libraries.

**Comment by fire** [comment]
What are your approaches?

I don't think adding rust SDKs are in the future.

**Comment by fire** [comment]
* https://github.com/xiph/rav1e
* `rav1e` currently requires Rust 1.51.0 or later to build.

* https://code.videolan.org/videolan/dav1d
* decoder only
* asm / c

**Comment by YakoYakoYokuYoku** [comment]
If you are speaking about bundling an AV1 codec, then the above can obviously turn one or more down, but my idea was more inclined into having a generic API for any AV1 library used as a backend, compare with [`python-qtpy`](https://github.com/spyder-ide/qtpy), which uses either `python-qt5` or `python-pyside2` to provide Qt for Python.

**Comment by fire** [comment]
Godot Engine uses a bundled approach. I personally call it the "I can compile offline" -- approach.

**Comment by novomesk** [comment]
Here is the size of libavif and libaom on Linux:

```
   149320 /usr/lib64/libavif.so.11.0.0
  6090936 /usr/lib64/libaom.so.3.1.0
```

As you can see the size of libavif alone is small, I think it would be waste of effort to implement another parser and convertor between YUV and RGB formats instead of it.

Size of rav1e on my Linux seems significantly bigger:

```
129687426 /usr/lib64/librav1e.a
 32496752 /usr/lib64/librav1e.so.0.4.1
```

I really recommend libavif+libaom. You can do local/static builds easily, they are portable. I personally tried Linux (amd64, power PC), Windows (mingw under MSYS2, MSVC) and MacOS.
libaom is made by Google. libavif made by Netllix but Google reviews it deeply and also contribute there (because libavif is included in Google Chrome).

dav1d is good and fast decoder, its 8bit decoding performance is higher than libaom, but I observed that even libaom alone is fast enough for AVIF.

It depends how much space(megabytes) you allow for the new feature.

**Comment by Calinou** [comment]
We discussed this proposal in a meeting, and the consensus is that this can be implemented with a GDExtension for now. As of July 2022, AVIF usage on the web is still not that widespread (even compared to other "modern" formats like WebP). There's also some uncertainty as to whether AVIF or JPEG XL will "win the battle" of the next-gen image format. It's *possible* that both formats end up coexisting with significant usage on the web, but I don't see it happening personally.

If AVIF adoption increases *significantly* in the future, we can reconsider this proposal and make AVIF support a core feature. (This also applies to JPEG XL.)

**Comment by YakoYakoYokuYoku** [comment]
ACK and thanks for discussing it. At the time I didn't saw the possibility of doing it in a GDExtension and I thought that this could only be done in core but if someone were to work on it (or it's actually) then I'd be pleased to hear. With that said I have no more comments.
